corehq/apps/app_manager/suite_xml/post_process/resources.py

```diff
@@ -0,0 +1,66 @@
+from collections import Counter
+from django.db import models
+
+from corehq.apps.app_manager.exceptions import ResourceOverrideError
+from corehq.apps.app_manager.suite_xml.contributors import PostProcessor
+from corehq.apps.app_manager.suite_xml.sections.resources import FormResourceContributor
+from corehq.apps.app_manager.suite_xml.xml_models import XFormResource
+
+
+class ResourceOverride(models.Model):
+    domain = models.CharField(max_length=255, null=False)
+    app_id = models.CharField(max_length=255, null=False)
+    # Type of resource, e.g., xform. Populated by the root_name of the relevant suite_xml.xml_models class.
+    root_name = models.CharField(max_length=32, null=False)
+    pre_id = models.CharField(max_length=255, null=False)
+    post_id = models.CharField(max_length=255, null=False)
+
+    class Meta(object):
+        unique_together = ('domain', 'app_id', 'root_name', 'pre_id')
+        index_together = ('domain', 'app_id', 'root_name')
```

Comment:
This `index_together` is not useful. these three columns are the leading columns of the `unique_together` so any query that would use these index could use the unique index instead

Reply:
I'm unfamiliar with that technique - is that supported in the Django ORM?  How does one query the unique index?  (Or does postgres optimize that type of query automagically?)

Reply:
There isn't any "technique" here. Unique constraints in postgres are implemented via an index as noted in https://www.postgresql.org/docs/9.6/indexes-unique.html

> PostgreSQL automatically creates a unique index when a unique constraint or primary key is defined for a table

And then as noted in the docs on [multicolumn indexes](https://www.postgresql.org/docs/9.6/indexes-multicolumn.html), an index can technically be used when a query queries any subset of the columns, but is most effective when there are equality constraints on the leftmost columns

> The exact rule is that equality constraints on leading columns, plus any inequality constraints on the first column that does not have an equality constraint, will be used to limit the portion of the index that is scanned

If the order of these columns were ordered differently from each other, then its possible that they could be useful, but because the multicolumn index has the same order as the unique index, then  the unique index is (almost) strictly better in all cases. 

Of course this is all ignoring the fact that this table may very well be small enough that postgres never actually uses the index and performs a sequential scan anyways. This can happen when the table is very small (so overhead of loading the entire table in memory is small) or that query is expected to return a large portion of the table (say on ICDS where there's only 2 or 3 real apps.

Reply:
Nifty, thanks for the explanation!